Source/DafnyCore/Resolver.cs

```diff
@@ -9941,6 +9944,51 @@ public CheckDividedConstructorInit_Visitor(Resolver resolver)
       return true;
     }
 
+    bool CheckIfEqualityIsDefinitelyNotSupported(IndDatatypeDecl dt, Graph<IndDatatypeDecl/*!*/>/*!*/ dependencies) {
+
+      var scc = dependencies.GetSCC(dt);
+
+      foreach (var ctor in dt.Ctors) {
+
+        if (ctor.IsGhost) {
+          return true;
+        }
+
+        foreach (var arg in ctor.Formals) {
+          var anotherIndDt = arg.Type.AsIndDatatype;
+          if (arg.IsGhost ||
+              (anotherIndDt != null && anotherIndDt.EqualitySupport == IndDatatypeDecl.ES.Never) ||
+              arg.Type.IsCoDatatype ||
+              arg.Type.IsArrowType ||
+              (arg.Type.IsOpaqueType && !arg.Type.SupportsEquality)) {
```

Comment:
What does `arg.Type.SupportsEquality` say for codatatypes and arrow types?  If it already says `false`, could we simplify the above to 
```suggestion
              !arg.Type.SupportsEquality) {
```
?

Reply:
That's what I did, initially, but this turns out to be incorrect, I think. The reason being that some types are initialized with SupportsEquality set to false, and because of recursive definitions, this Boolean may not have been resolved yet. Thoughts?

Reply:
I also like that by writing the code this way, even if it might be less efficient, the code reads more like its specification of what it should do.